evm/src/bn254.rs

```diff
@@ -0,0 +1,743 @@
+        (0, 0),
+        (0, 0),
+        (0, 0),
+        (0, 1),
+        (1, 0),
+        (1, 1),
+        (0, 1),
+        (1, 1),
+        (1, 0),
+        (0, 1),
+        (0, 0),
+        (1, 0),
+        (0, 1),
+        (1, 0),
+        (1, 1),
+        (1, 0),
+        (1, 1),
+        (0, 1),
+        (1, 1),
+    ];
+
+    const EXPS0: [usize; 65] = [
+        0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0,
+        0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1,
+        0, 0, 1, 1, 0,
+    ];
+
+    let mut sq: Fp12 = f;
+    let mut y0: Fp12 = embed_fp12(U256::one());
+    let mut y2: Fp12 = embed_fp12(U256::one());
+    let mut y4: Fp12 = embed_fp12(U256::one());
+
+    for (a, b, c) in EXPS4 {
+        if a != 0 {
+            y4 = mul_fp12(y4, sq);
+        }
+        if b != 0 {
+            y2 = mul_fp12(y2, sq);
+        }
+        if c != 0 {
+            y0 = mul_fp12(y0, sq);
+        }
+        sq = mul_fp12(sq, sq);
+    }
+    y4 = mul_fp12(y4, sq);
+
+    for (a, b) in EXPS2 {
+        if a != 0 {
+            y2 = mul_fp12(y2, sq);
+        }
+        if b != 0 {
+            y0 = mul_fp12(y0, sq);
+        }
+        sq = mul_fp12(sq, sq);
+    }
+    y2 = mul_fp12(y2, sq);
+
+    for a in EXPS0 {
+        if a != 0 {
+            y0 = mul_fp12(y0, sq);
+        }
+        sq = mul_fp12(sq, sq);
+    }
+    y0 = mul_fp12(y0, sq);
+
+    y0 = inv_fp12(y0);
+
+    y4 = mul_fp12(y4, y2);
+    y4 = mul_fp12(y4, y2);
+    y4 = mul_fp12(y4, y0);
+
+    y4 = frob_fp12(1, y4);
+    y2 = frob_fp12(2, y2);
+
+    mul_fp12(mul_fp12(y4, y2), y0)
+}
+
+pub fn tangent(p: Curve, q: TwistedCurve) -> Fp12 {
+    let [px, py] = p;
+    let [qx, qy] = q;
+
+    let cx = neg_fp(mul_fp(U256::from(3), mul_fp(px, px)));
+    let cy = mul_fp(U256::from(2), py);
+
+    sparse_embed(
+        sub_fp(mul_fp(py, py), U256::from(9)),
+        mul_fp2(embed_fp2(cx), qx),
+        mul_fp2(embed_fp2(cy), qy),
+    )
+}
+
+pub fn cord(p1: Curve, p2: Curve, q: TwistedCurve) -> Fp12 {
+    let [p1x, p1y] = p1;
+    let [p2x, p2y] = p2;
+    let [qx, qy] = q;
+
+    let cx = sub_fp(p2y, p1y);
+    let cy = sub_fp(p1x, p2x);
+
+    sparse_embed(
+        sub_fp(mul_fp(p1y, p2x), mul_fp(p2y, p1x)),
+        mul_fp2(embed_fp2(cx), qx),
+        mul_fp2(embed_fp2(cy), qy),
+    )
+}
+
+fn tangent_slope(p: Curve) -> Fp {
+    let [px, py] = p;
+    let num = mul_fp(mul_fp(px, px), U256::from(3));
+    let denom = mul_fp(py, U256::from(2));
+    div_fp(num, denom)
+}
+
+fn cord_slope(p: Curve, q: Curve) -> Fp {
+    let [px, py] = p;
+    let [qx, qy] = q;
+    let num = sub_fp(qy, py);
+    let denom = sub_fp(qx, px);
+    div_fp(num, denom)
+}
+
+fn third_point(m: Fp, p: Curve, q: Curve) -> Curve {
+    let [px, py] = p;
+    let [qx, _] = q;
+    let ox = sub_fp(mul_fp(m, m), add_fp(px, qx));
+    let oy = sub_fp(mul_fp(m, sub_fp(px, ox)), py);
+    [ox, oy]
+}
+
```

Comment:
I know these are standard elliptic curve functions, but there should still be a link to the source or some other kind of spec.